### PR TITLE
ObservableSnapshotArray

### PR DESCRIPTION
**Pull request by samtstern**
Changes made:
  * Define `ObservableSnapshotArray` base class which is what adapters expect
  * `FirebaseArray` is an implementation of `ObservableSnapshotArray`
  * `FirebaseIndexArray` **contains** a `FirebaseArray` rather than being one

Benefits:
  * Code is much DRY-er
  * `FirebaseIndexArray` reads much cleaner, does not contain the hacks it did before
  * Remove `FirebaseArrayOfObjects` class (and subclasses) since they are not necessary

You previously asked the question as to why `OSA` is a `List<DataSnapshot>` and not a `List<Object>`, here is my rationale:
  * `OSA` should stick to what developers find hard: making RTDB data feel like a `List`.  This is achieved by exposing `DataSnapshot`.  Converting to `Object` is pretty easy and they may want to customize that behavior.
  * Developers frequently need access to metadata (`DataSnapshot`) so this will have to be tracked and exposed either way

The following questions remain:
  * Naming: is there a clearer name than `ObservableSnapshotArray`?
  * Breaking changes: what are the 1.0-->2.0 breaking changes caused by this change?  Are they any more avoidable now?
  * What is the purpose of the `FirebaseAdapter` interface?  We are not consuming it anywhere, so can we remove it?

### CODE REVIEW

File: database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java [hunk 1]
```diff
@@ -146,17 +179,21 @@ public void onDataChange(DataSnapshot snapshot) {
...
                 }
             } else {
                 if (isKeyAtIndex(key, index)) {
-                    mDataSnapshots.remove(index);
-                    notifyChangeEventListeners(ChangeEventListener.EventType.REMOVED, index);
+                    // This data has disappeared, remove it
+                    removeData(index);
+                    notifyChangeEventListeners(ChangeEventListener.EventType.REMOVED, snapshot, index);
                 } else {
+                    // Data we never knew about has disappeared
```
Comment (SUPERCILEX) [inline database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java]: This is technically incorrect: it didn't disappear, the ref just doesn't exist in the RTDB.

File: database/src/main/java/com/firebase/ui/database/FirebaseListAdapter.java [hunk 2]
```diff
@@ -57,13 +57,18 @@ public FirebaseListAdapter(Activity activity,
...
                                Class<T> modelClass,
                                @LayoutRes int modelLayout,
                                Query query) {
-        this(activity, new FirebaseArray(query), modelClass, modelLayout);
+        mSnapshots = new FirebaseArray<T>(query, this);
+        mActivity = activity;
+        mModelClass = modelClass;
+        mLayout = modelLayout;
+
+        startListening();
```
Comment (SUPERCILEX) [inline database/src/main/java/com/firebase/ui/database/FirebaseListAdapter.java]: I feel like maintaining two basically identical constructor will be a pain so could we could do this instead:
```java
protected void init(Activity activity,
                    ObservableSnapshotArray<T> snapshots,
                    Class<T> modelClass,
                    @LayoutRes int modelLayout) {
    mSnapshots = snapshots;
    mActivity = activity;
    mModelClass = modelClass;
    mLayout = modelLayout;

    startListening();
}
```

File: database/src/main/java/com/firebase/ui/database/FirebaseRecyclerAdapter.java [hunk 3]
```diff
@@ -62,13 +63,18 @@ public FirebaseRecyclerAdapter(FirebaseArray snapshots,
...
      *              {@code endAt()}.
-     * @see #FirebaseRecyclerAdapter(FirebaseArray, Class, int, Class)
+     * @see #FirebaseRecyclerAdapter(ObservableSnapshotArray, Class, int, Class)
      */
     public FirebaseRecyclerAdapter(Class<T> modelClass,
                                    @LayoutRes int modelLayout,
                                    Class<VH> viewHolderClass,
                                    Query query) {
-        this(new FirebaseArray(query), modelClass, modelLayout, viewHolderClass);
+        mSnapshots = new FirebaseArray<T>(query, this);
```
Comment (SUPERCILEX) [inline database/src/main/java/com/firebase/ui/database/FirebaseRecyclerAdapter.java]: Same as above, a `Ctrl + Alt + M` on the constructor above this one should simplify this.

File: database/src/main/java/com/firebase/ui/database/ObservableSnapshotArray.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+ *
+ * @param <T> a POJO class to which the DataSnapshots can be converted.
+ */
+public abstract class ObservableSnapshotArray<T> extends ImmutableList<DataSnapshot> {
+
+    protected final List<ChangeEventListener> mListeners = new CopyOnWriteArrayList<>();
+    protected final SnapshotParser<T> mParser;
+
+    public ObservableSnapshotArray() {
+        mParser = null;
```
Comment (samtstern) [inline database/src/main/java/com/firebase/ui/database/ObservableSnapshotArray.java]: Ok I added this type of constructor to each class.

**Comment by samtstern** [comment]
Addressed all the code-level comments.  The two remaining items are larger (`FirebaseAdapter` and `FirebaseArrayOfObjects`):

  * For `FirebaseAdapter` I am concerned about introducing an interface that we don't use since it just restricts us to a certain signature without a clearly defined use case.
  * Is `FirebaseArrayOfObjects` something we could introduce in the future without breaking anything?  Again I feel like adding code we don't use at all makes it hard to know if we're making something useful.

File: database/src/main/java/com/firebase/ui/database/CachingObservableSnapshotArray.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        super(parser);
+    }
+
+    @Override
+    public T getObject(int index) {
+        String key = get(index).getKey();
+
+        // Return from the cache if possible, otherwise populate the cache and return
+        if (mObjectCache.containsKey(key)) {
+            return mObjectCache.get(key);
```
Comment (samtstern) [inline database/src/main/java/com/firebase/ui/database/CachingObservableSnapshotArray.java]: I think `null` is probably valid as a return value in some custom cases.

Comment (SUPERCILEX) [inline database/src/main/java/com/firebase/ui/database/CachingObservableSnapshotArray.java]: @samtstern SGTM!

**Comment by SUPERCILEX** [comment]
@samtstern For `FirebaseAdapter`, we could make it package-private thus giving us the freedom to change it however we like. It could look something like this:
```java
/**
 * A helper interface to ensure adapter consistency and simplify documentation.
 */
interface FirebaseAdapter<T> extends ChangeEventListener, SnapshotParser<T> {...}
```

For `FirebaseArrayOfObjects`, I think it would work fine to introduce it in the future if there's popular demand so I agree with you on not including it right now. I keep forgetting that I can use `FirebaseArray` in my own code now! 😄

@samtstern Are we good to merge after `FirebaseAdapter` is resolved?

**Comment by samtstern** [comment]
@SUPERCILEX I made `FirebaseAdapter` non-public, I think that's a good compromise.  Let's merge this into your branch and then hopefully we should get the big kahuna merged very soon!

The last thing I want to do is have you go through your test app again and make an explicit list of the breaking changes we're making (since I lost mental track of that many commits ago)

**Comment by SUPERCILEX** [comment]
@samtstern SGTM! I started merging this PR on my machine and adding my cleanup commit, but I noticed that we [weren't using the `index` variable in `FirebaseIndexArray`](https://github.com/SUPERCILEX/FirebaseUI-Android/pull/24/files#diff-848b2a3fc375b8ab8b5f22f22157614aR177). The tests pass both ways (using `index` instead of `newIndex` and vice versa), but I wanted to make sure this wasn't intentional on your part before removing `newIndex` and using `index` instead.

**Comment by samtstern** [comment]
@SUPERCILEX I mostly didn't change to use `index` since the tests passed like you said, and I was trying not to break anything.  Fine with me either way you want to do it!

**Comment by SUPERCILEX** [comment]
@samtstern Ok, I went with using `index` instead of `newIndex`. Merging!

**Pull request merged**